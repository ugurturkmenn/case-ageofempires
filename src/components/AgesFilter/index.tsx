import { useDispatch } from 'react-redux';

const AgesFilter = () => {
  const ages = [
    {
      id: 1,
      name: "ALL",
      value: "ALL"
    },
    {
      id: 2,
      name: "DARK",
      value: "DARK"
    },
    {
      id: 3,
      name: "FEUDAL",
      value: "FEUDAL"
    },
    {
      id: 4,
      name: "CASTLE",
      value: "CASTLE"
    },
    {
      id: 5,
      name: "IMPERIAL",
      value: "IMPERIAL"
    }
  ];

  const dispatch = useDispatch();

  const handleAge = (e: React.MouseEvent<HTMLButtonElement>) => {
    const val = (e.target as HTMLButtonElement).value
    dispatch({ type: "AGE_FILTER", payload: val });
  }

  return (
    <div data-testid="ages-filter">
      <h2>Ages</h2>
      <div className='flex gap-3 mb-3 mt-3 text-white'>
        {ages && ages.map((age) => (<button key={age?.id} className='bg-slate-500 p-1' onClick={handleAge} value={age.value}>
          {age?.name}</button>))}
      </div>
    </div>
  )
}

export default AgesFilter
