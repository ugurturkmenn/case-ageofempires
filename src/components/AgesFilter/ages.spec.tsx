import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import AgesFilter from ".";

const AGE_FILTER = 'AGE_FILTER';

const ageFilter = (payload: any) => ({
  type: AGE_FILTER,
  payload
});

const initialState = {
  units: {
    units: [{
      id: 1,
      name: "Archer",
      description: "Quick and light. Weak at close range; excels at battle from distance",
      expansion: "Age of Kings",
      age: "Feudal",
      cost: { Wood: 25, Gold: 45 },
      build_time: 35,
      reload_time: 2,
      attack_delay: 0.35,
      movement_rate: 0.96,
      line_of_sight: 6,
      hit_points: 4,
      range: 30,
      attack: 4,
      armor: "0/0",
      accuracy: "80%"
    }],
    filteredUnits: [{
      id: 1,
      name: "Archer",
      description: "Quick and light. Weak at close range; excels at battle from distance",
      expansion: "Age of Kings",
      age: "Feudal",
      cost: { Wood: 25, Gold: 45 },
      build_time: 35,
      reload_time: 2,
      attack_delay: 0.35,
      movement_rate: 0.96,
      line_of_sight: 6,
      hit_points: 4,
      range: 30,
      attack: 4,
      armor: "0/0",
      accuracy: "80%"
    }],
    age: 'ALL',
    loading: false,
    error: null,
    costFilters: {
      Gold: { checked: false, range: { min: 0, max: 200 } },
      Food: { checked: false, range: { min: 0, max: 200 } },
      Wood: { checked: false, range: { min: 0, max: 200 } },
    }
  }
};

const mockStore = configureStore();

const renderWithStore = (ui: any, { store }: any) => {
  return render(<Provider store={store}>{ui}</Provider>);
};

describe('AgesFilter Component', () => {

  test('should match the snapshot', () => {
    const store = mockStore(initialState);
    const { asFragment } = renderWithStore(<AgesFilter />, { store });
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render all age filter buttons', () => {
    const store = mockStore(initialState);
    renderWithStore(<AgesFilter />, { store });

    expect(screen.getByText(/ALL/i)).toBeInTheDocument();
    expect(screen.getByText(/DARK/i)).toBeInTheDocument();
    expect(screen.getByText(/FEUDAL/i)).toBeInTheDocument();
    expect(screen.getByText(/CASTLE/i)).toBeInTheDocument();
    expect(screen.getByText(/IMPERIAL/i)).toBeInTheDocument();
  });

  test('should dispatch correct action on button click', () => {
    const store = mockStore(initialState);
    store.dispatch = jest.fn();

    renderWithStore(<AgesFilter />, { store });

    const allButton = screen.getByText(/ALL/i);
    fireEvent.click(allButton);
    expect(store.dispatch).toHaveBeenCalledWith(ageFilter('ALL'));

    const darkButton = screen.getByText(/DARK/i);
    fireEvent.click(darkButton);
    expect(store.dispatch).toHaveBeenCalledWith(ageFilter('DARK'));

  });
});
