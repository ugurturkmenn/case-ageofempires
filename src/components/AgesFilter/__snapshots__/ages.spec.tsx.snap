// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgesFilter Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="ages-filter"
  >
    <h2>
      Ages
    </h2>
    <div
      class="flex gap-3 mb-3 mt-3 text-white"
    >
      <button
        class="bg-slate-500 p-1"
        value="ALL"
      >
        ALL
      </button>
      <button
        class="bg-slate-500 p-1"
        value="DARK"
      >
        DARK
      </button>
      <button
        class="bg-slate-500 p-1"
        value="FEUDAL"
      >
        FEUDAL
      </button>
      <button
        class="bg-slate-500 p-1"
        value="CASTLE"
      >
        CASTLE
      </button>
      <button
        class="bg-slate-500 p-1"
        value="IMPERIAL"
      >
        IMPERIAL
      </button>
    </div>
  </div>
</DocumentFragment>
`;
